share/p2p/shrexeds/client.go

```diff
@@ -70,30 +73,64 @@ func (c *Client) RequestEDS(
 	return nil, err
 }
 
-func (c *Client) doRequest(
+type result struct {
+	eds *rsmt2d.ExtendedDataSquare
+	err error
+}
+
+func (c *Client) doSafeRequest(
 	ctx context.Context,
 	dataHash share.DataHash,
 	to peer.ID,
 ) (*rsmt2d.ExtendedDataSquare, error) {
+	resultChan := make(chan result)
 	stream, err := c.host.NewStream(ctx, to, c.protocolID)
 	if err != nil {
 		return nil, fmt.Errorf("failed to open stream: %w", err)
 	}
 
+	// if there is a deadline set, safetyContext will give the request 500 milliseconds to recognize the context
+	// has deadlined before cancelling the operation itself.
+	// this is a hotfix until we solve doRequests disrepect of context. Related: issue #2109
+	safetyContext, cancel := context.WithCancel(ctx)
 	if dl, ok := ctx.Deadline(); ok {
+		safetyContext, cancel = context.WithDeadline(context.Background(), dl.Add(safetyTimeout))
 		if err = stream.SetDeadline(dl); err != nil {
```

Comment:
can we actually just set the stream deadline regardless of whether ctx.Deadline contains a deadline or not? if no deadline contained, set to some min?

Reply:
This is what vlads pr handles